### PR TITLE
fix reset.yml for lxc containers, add support for extra worker packages

### PR DESCRIPTION
**Pull request by acdoussan**
# Proposed Changes

- implements https://github.com/techno-tim/k3s-ansible/discussions/259
- fixes the reset script for lxc containers, ref: https://github.com/techno-tim/k3s-ansible/pull/263

Tested by deploying to a test cluster with the following extra packages, and verifying they were present by using `apt list --installed | grep <package name>`. Also ran the reset script and verified that they were uninstalled using the same method.

```
extra_worker_packages:
- open-iscsi  # for longhorn
- nfs-common  # for longhorn backups

```

## Checklist

- [x] Tested locally
- [x] Ran `site.yml` playbook
- [x] Ran `reset.yml` playbook
- [x] Did not add any unnecessary changes
- [x] Ran pre-commit install at least once before committing
- [x] 🚀

### CODE REVIEW

File: reset.yml [hunk 1]
```diff
@@ -17,7 +17,6 @@
 - hosts: proxmox
   gather_facts: true
   become: yes
-  remote_user: "{{ proxmox_lxc_ssh_user }}"
```
Comment (timothystewart6) [inline reset.yml:20]: why was this removed?

**Comment by timothystewart6** [comment]
Thank you.  I think this is a good middle ground to allow someone to install a list of packages without too many changes or dependencies.    Why stop at just workers?  What about servers?  

Also, it would be awesome to write a molecule test for this.  Install the package and verify it's there.

File: reset.yml (hunk 1 above)
Comment (acdoussan) [reply reset.yml:20]: It is my understanding this doesn't do anything, per the other CR. We removed this from site.yml, so I updated this to match. I also removed my definition for proxmox_lxc_ssh_user because the root default is fine, and this caused the reset to fail because this variable wasn't defined.

https://github.com/techno-tim/k3s-ansible/pull/263

**Comment by acdoussan** [comment]
Whatcha mean by servers? Like the proxmox servers?

The nodes are VMs / containers, I don't know if it makes sense to define what packages you want on your proxmox servers within this playbook. This just lets you add packages to the worker nodes if you know you need them for your applications that will be running on the nodes.

I did briefly think about molecule, but it looks like the tests use the sample inventory, wasn't sure about the sample actually installing packages. Would you have a preferred place / way to set this only for a test? Or do you think it's fine for the sample to have something that would actually get installed?

**Comment by acdoussan** [comment]
oh, this is probably what the overrides file is for?

**Comment by acdoussan** [comment]
not sure if that update does anything or not or if it is the reason the default is failing, was trying to avoid downloading all of the stuff and just let the CI handle it but might have to

**Comment by timothystewart6** [comment]
This is failing because the package isn't available on rocky, only Ubuntu so you'll have to modify the tests to only install on `node1`, not `node2`.

Also, when I mentioned the servers, I meant the `master` nodes since I think this is going to be the next things someone asks for.

File: inventory/sample/group_vars/all.yml [hunk 2]
```diff
@@ -40,6 +40,10 @@ extra_server_args: >-
 extra_agent_args: >-
   {{ extra_args }}
 
+extra_worker_packages: []
+# - open-iscsi  # for longhorn
+# - nfs-common  # for longhorn backups
+
```
Comment (sleiner) [inline inventory/sample/group_vars/all.yml:46]: This variable are instantiated one time for every Ansible host, so I think a better approach to this feature would be to have an `extra_packages` variable that is set for specific hosts or groups (in this case, the equivalent would be the `node` group). That would give you a lot more flexibility (e.g. when wanting to install packages on control nodes, dealing with different distros, etc.)

File: molecule/default/tests/test_default_node.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    os.environ['MOLECULE_INVENTORY_FILE']).get_hosts('node')
+
+
+@pytest.mark.parametrize('pkg', [
+    'open-iscsi',
+    'nfs-common',
+])
+def test_pkg(host, pkg):
+    package = host.package(pkg)
+    assert package.is_installed
```
Comment (sleiner) [inline molecule/default/tests/test_default_node.py:16]: Up until now, we have had neither scenario-specific test cases (all current test cases are in `molecule/resources/verify/from_outside`) nor any test cases using `testinfra`. Can you explain why you want to introduce both concepts for this case?

File: roles/k3s/node/tasks/main.yml [hunk 4]
```diff
@@ -14,3 +14,8 @@
     daemon_reload: yes
     state: restarted
     enabled: yes
+
+- name: install extra worker node packages
```
Comment (sleiner) [inline roles/k3s/node/tasks/main.yml:18]:
```suggestion
- name: Install extra worker node packages
```

File: roles/reset/tasks/main.yml [hunk 5]
```diff
@@ -82,3 +82,8 @@
     path: /etc/rc.local
     state: absent
   when: proxmox_lxc_configure and rcfile.stat.exists and ((rcslurp.content | b64decode).splitlines() | length) <= 1
+
+- name: remove extra worker node packages
```
Comment (sleiner) [inline roles/reset/tasks/main.yml:86]:
```suggestion
- name: Remove extra worker node packages
```

Comment (sleiner) [inline roles/reset/tasks/main.yml:86]: This should not affect non-worker hosts, but it does.

**Review by sleiner — COMMENTED** [review]
I have reviewed this under the premise that installing packages unrelated to k3s is a desirable feature (which I still disagree with, but that's a different discussion and not for me to decide).

**Comment by acdoussan** [comment]
> This is failing because the package isn't available on rocky, only Ubuntu so you'll have to modify the tests to only install on `node1`, not `node2`.
> 
> Also, when I mentioned the servers, I meant the `master` nodes since I think this is going to be the next things someone asks for.

Oh gotcha, no user workloads run on the control nodes, right? I would think that any packages that need to be installed on the control nodes are the responsibility of the playbook, not the user. I do like the idea @sleiner proposed though with `extra_packages` but I think this depends on the mindset of allowing this.

My thought process is that the playbook is setting up a k3s cluster to run my applications, and my applications need X packages to run, so it's the responsibility of the playbook to make sure these are installed. Making sure that other packages are installed so that I can use the hosts for more than just k3s isn't really in this scope. I'm not sure we really want to offer to install extra packages on the control nodes, but admittedly it would probably be relatively easy to do it for any host, so I'm happy to take either approach.

File: molecule/default/tests/test_default_node.py (hunk 3 above)
Comment (acdoussan) [reply molecule/default/tests/test_default_node.py:16]: In short, because I was using the following example and my molecule knowledge is pretty limited. Do you have a recommended alternative way to implement this?

File: roles/k3s/node/tasks/main.yml (hunk 4 above)
Comment (acdoussan) [reply roles/k3s/node/tasks/main.yml:18]: will take

File: roles/reset/tasks/main.yml (hunk 5 above)
Comment (acdoussan) [reply roles/reset/tasks/main.yml:86]: will take

Comment (acdoussan) [reply roles/reset/tasks/main.yml:86]: That's a good point, assuming we keep this limited to worker nodes, should there be a new reset role for worker nodes that has this in it? Or something else?

File: inventory/sample/group_vars/all.yml (hunk 2 above)
Comment (acdoussan) [reply inventory/sample/group_vars/all.yml:46]: Is this as easy as defining another var file with the right name? Think this is an interesting approach, just not totally sure what it would look like

**Comment by timothystewart6** [comment]
Hey, sorry but I am going to close this PR, I should have when it was opened however the more I think about it the more it doesn't make a lot of sense for this playbook to take on the dependencies of something that's not required to run k3s.  While there might be features and changes that are seemingly easy to add, at the end of the day it introduces complexity to this playbook that is either unnecessary or untested.  Thank you again for the suggestion and thank you @sleiner for putting in time to do a code review even though this isn't a potential merge candidate.  Thank for understanding.

**Comment by acdoussan** [comment]
No problem! Sorry for going MIA for a while, work has been consuming most of my tech energy recently.

I've also been thinking about this and I think the real problem is that the playbook is the "interface" rather than the roles. This means if you want to do anything custom, you have to diverge from the upstream and figure out how to pull down changes while also keeping your changes. Not a huge deal with a separate branch and rebases, but can lead to merge conflicts and just general headaches.

gerlingguy has a sort of similar mac dev playbook, but the roles there are the "interface". This is accomplished by publishing the roles to ansible galaxy as a collection, and instead of forking his playbook, there is a link to a repo template to get you started. This way you still get all of the setup like forking, but you actually own the repo, and can just add tasks like any other playbook without worry about diverging. Updates are pulled by pulling the new role versions.

I think there is already an open issue for this (https://github.com/techno-tim/k3s-ansible/issues/198) so this is probably where effort should be focused to allow this 🙂

**Comment by acdoussan** [comment]
Added some closing thoughts to the discussion if anyone comes along and is wondering what their options are.

Discussion can be found here: https://github.com/techno-tim/k3s-ansible/discussions/259

**Comment by sleiner** [comment]
@acdoussan I just saw that there were some open review questions for me that I forgot about - sorry about that :(
